### DISCUSSION TITLE
Prefix in Screen Variants naming

### DISCUSSION THREAD

**Discussion by lamualfa** [windicss/windicss#28] [Ideas]
## Current

Variant | Rule | Description
-- | -- | --
sm | @media (min-width: 640px) { ... } | Enable utility when the screen width is greater than 640px
-sm | @media (max-width: 640px) { ... } | Enable utility when the screen width is less than 640px
+sm | @media (min-width: 640px) and (max-width: 768px) { ... } | Enable utility when the screen width is greater than 640px and less than 768px

## New Ideas

Variant | Rule | Description
-- | -- | --
+sm | @media (min-width: 640px) { ... } | Enable utility when the screen width is greater than 640px
-sm | @media (max-width: 640px) { ... } | Enable utility when the screen width is less than 640px
sm | @media (min-width: 640px) and (max-width: 768px) { ... } | Enable utility when the screen width is greater than 640px and less than 768px

### Description

- Use `+` for greater than 
- Use `-` for less than 
- Not using anything for range

## Another Ideas

Variant | Rule | Description
-- | -- | --
\>sm | @media (min-width: 640px) { ... } | Enable utility when the screen width is greater than 640px
<sm | @media (max-width: 640px) { ... } | Enable utility when the screen width is less than 640px
sm | @media (min-width: 640px) and (max-width: 768px) { ... } | Enable utility when the screen width is greater than 640px and less than 768px

### Description

- Use `>` for greater than. Equal to `screen > sm`
- Use `<` for less than. Equal to `screen < sm`
- Not using anything for range

<hr>

I know it will change the standards in Tailwind. But what do you think if WindiCSS uses it in a future version? let's find out what you think about the idea above?

**Comment by voorjaar** [comment]
I want to know about why you tend to use `>` and `<`.

For me,

`+` means only works on that screen.

`-` represents the opposite direction to mobile first.

**Reply by voorjaar** [reply]
But I think we can add a configuration to allow you to modify this option. Is this okay for you?

**Reply by lamualfa** [reply]
Yes. That is what I mean.

**Reply by lamualfa** [reply]
> I want to know about why you tend to use `>` and `<`.
> 
> For me,
> 
> `+` means only works on that screen.
> 
> `-` represents the opposite direction to mobile first.

<hr/>

Do not have to use the `<` or the `>` sign. `-` and `+` signs are good enough. I'm just trying to apply the semantic naming concept to the class name that will be used. 

**Examples of cases of using the `+` symbol in school grades.**

- `A` for the value of A standard.
- `A+` for a more value. Greater than `A` standard.
- `A-` for the value of A. but not higher than the `A` standard.

![image](https://user-images.githubusercontent.com/39755201/108026525-131dff80-705b-11eb-8663-5ab603c43989.png)

> Ref: https://en.wikipedia.org/wiki/Academic_grading_in_the_United_States

<hr/>

If applied to the Screen Variant namings, it looks like this:


Variant | Rule | Description
-- | -- | --
+sm | @media (min-width: 640px) { ... } | Enable utility when the screen width is greater than 640px
sm | @media (min-width: 640px) and (max-width: 768px) { ... } | Enable utility when the screen width is greater than 640px and less than 768px
-sm | @media (max-width: 640px) { ... } | Enable utility when the screen width is less than 640px

**Reply by ElMassimo** [reply]
The value of a utility-first library is consistency, so the defaults should match the ones in Tailwind CSS.

In that regard, `sm` should only use `min-width`, because that's what people have come to expect from Tailwind CSS.

<hr/>

I do agree that `+sm` might be too ambiguous, we could use a different prefix for that [range variant](https://windicss.netlify.app/utilities/variants.html#only-screen), such as [`only:sm`](https://windicss.netlify.app/utilities/variants.html#only-screen) or even a suffix  such as [`sm-only`](https://windicss.netlify.app/utilities/variants.html#only-screen).

**Reply by voorjaar** [reply]
But it's a bit long, to be honest.

**Reply by ElMassimo** [reply]
Ranges will probably be used less often than the other two options, so class length it's less of a concern, and [`only`](https://windicss.netlify.app/utilities/variants.html#only-screen) does match the [mental model](https://windicss.netlify.app/utilities/variants.html#only-screen) for this feature.

**Reply by lamualfa** [reply]
Use `~` as a prefix instead `only` prefix or `+` symbol looks like it could be a good alternative.

**Example**

- `~sm:w-1`
- `~md:align-middle`
- `~lg:bg-gray-300`

Let's find out what do you think? @voorjaar @ElMassimo

**Reply by voorjaar** [reply]
> Use `~` as a prefix instead `only` prefix or `+` symbol looks like it could be a good alternative.
> 
> **Example**
> 
> * `~sm:w-1`
> * `~md:align-middle`
> * `~lg:bg-gray-300`
> 
> Let's find out what do you think? @voorjaar @ElMassimo

I like the idea of `~` prefix. But it is a bit unclear, when using `-sm:w-1 ~md:aligh-middle` at the same time. Could you think of another alternative to `-`?

**Reply by lamualfa** [reply]
What about equal `=` symbol?

**Example**

- `=sm:w-1`
- `=md:align-middle`
- `=lg:bg-gray-300`
- `-sm:w-1 =md:aligh-middle`

<hr/>

or another alternative is using `#` symbol.

**Example**

- `#sm:w-1`
- `#md:align-middle`
- `#lg:bg-gray-300`
- `-sm:w-1 #md:aligh-middle`

<hr/>

or we can follow the rule of writing important attribute that use `!`.

**Example**

- `!sm:w-1`
- `!md:align-middle`
- `!lg:bg-gray-300`
- `-sm:w-1 !md:aligh-middle`

**Reply by voorjaar** [reply]
How about this?
`sm` -> `@media (min-width: 640px)`
`~sm` -> `@media (max-width: 640px)`
`@sm` -> `@media (min-width: 640px) and (max-width: 768px) `

**Reply by lamualfa** [reply]
Yeah that is good. But I think it seems better if you stick to the `+` or `-` instead `~` sign for the max-width rule.

The reason is because in Math, the `~` symbol is used to compare similar number. And it's a little weird to use as a max-width rule.

![images (9).jpeg](https://user-images.githubusercontent.com/39755201/108324592-afcad380-71fa-11eb-80f7-e1b356f3e255.jpeg)

**Reply by ElMassimo** [reply]
I like the equals sign! Combining the proposals, adding `!`:

```css
sm -> @media (min-width: 640px)
!sm -> @media (max-width: 640px)
=sm -> @media (min-width: 640px) and (max-width: 768px)
```

**Reply by voorjaar** [reply]
I just thought we could use `!` To generate important style. Such as,

```css
!bg-red-500 {
   background: red important!;
}
```

**Reply by sastan** [reply]
@voorjaar I'm the maintainer of twind, which is a runtime tailwind solution. I just want to jump here: We are using `!` for important style too but as a suffix like `bg-red-500!`. Your library looks really great. Maybe we can talk about having similar syntax for these tailwind extension.  Here is a [list of our extensions](https://twind.dev/docs/handbook/getting-started/tailwind-extensions.html). I'm happy to chat any time you want.

**Reply by voorjaar** [reply]
Very interesting, I am willing to talk.

**Reply by lamualfa** [reply]
@sastan Hey Sastan. Currently i using Twind in my project. Great library. Thanks. But i still waiting Typography plugin that can be use with Twind in new version.

**Reply by voorjaar** [reply]
> @voorjaar I'm the maintainer of twind, which is a runtime tailwind solution. I just want to jump here: We are using `!` for important style too but as a suffix like `bg-red-500!`. Your library looks really great. Maybe we can talk about having similar syntax for these tailwind extension. Here is a [list of our extensions](https://twind.dev/docs/handbook/getting-started/tailwind-extensions.html). I'm happy to chat any time you want.

https://discord.gg/mUsrtzaV. Is discord chat is fine for you?

**Reply by sastan** [reply]
> @sastan Hey Sastan. Currently i using Twind in my project. Great library. Thanks. But i still waiting Typography plugin that can be use with Twind in new version.

Sorry for that. I have so many things to do. But I‘ll put near the top of my list.

**Reply by lamualfa** [reply]
> > @sastan Hey Sastan. Currently i using Twind in my project. Great library. Thanks. But i still waiting Typography plugin that can be use with Twind in new version.
> 
> Sorry for that. I have so many things to do. But I‘ll put near the top of my list.

Not problem. Don't be pressured. it's an open source project. We're not paid to solve it. so first prioritize jobs that can give you money for your daily needs.

**Reply by schelmo** [reply]
if using mobile first, wouldnt it be more practical ommiting 1px in max-width if using "+" or "-" ?
if you doing something like this:
class="-lg:bg-blue-100 lg:w-2"
both gets applied when the screen is exactly 1024px 
or would adding "<" and ">" making more sense?

**Reply by voorjaar** [reply]
Adding "<" and ">" making more sense? Yes.

**Reply by schelmo** [reply]
or a better example:
class="bg-white -lg:focus-within:bg-blue-100"
which is applied also at exactly 1024px.
you must overwrite it then for one pixel with: lg:focus-within:bg-white

**Reply by lamualfa** [reply]
@voorjaar 

Does this discussion need to be closed?

<hr>

@ElMassimo 

> The value of a utility-first library is consistency, so the defaults should match the ones in Tailwind CSS.


Tailwind just released a JIT version. Stop follow the Tailwind CSS. We have to make a different one 🔥🔥🔥.

Windi CSS != Tailwind CSS

**Reply by ahku** [reply]
LTTP: I think we should stick to the default being mobile-first, ie. `sm:block` means `min-width`. Then use the max and between variants as overrides. `-sm:`, `<sm:`, both work. I particularly like the `@sm:` as a between breakpoint.

Btw, not sure who said that "Windi CSS != Tailwind CSS" and that we should stop following, etc. It was probably deleted but I think we can all agree that this is not what we should strive for. The goal should be to to combine everybody's battle-tested experience into making the best tools available. That's what make Tailwind powerful and valuable, and that's what gonna make Windi CSS great. Being able to take a Tailwind project and turn it into a Windi-powered project with little effort _and_ gain new fantastic possibilities is so good. But of course, Windi can be even more, but that topic is better left for be another thread.

**Reply by ElMassimo** [reply]
Edited formatting in the previous comment to clarify which part was quote and which part was commentary.

I still think [consistency with the ecosystem](https://github.com/windicss/windicss/discussions/28#discussioncomment-374050) is important.

**Reply by ahku** [reply]
> Edited formatting in the previous comment to clarify which part was quote and which part was commentary.
> 
> 
> 
> I still think [consistency with the ecosystem](https://github.com/windicss/windicss/discussions/28#discussioncomment-374050) is important.

Ah now I see. That was OP's opinion. I agree with conforming to the ecosystem as much as possible. Not for my own sake, but because the value of standards means than developers can more be productive across code bases.
